doc/developer/design/20230418_persist_push_based_updates.md

```diff
@@ -0,0 +1,278 @@
+
+As Materialize is currently architected, Persist serves as a type of communication layer between processes. A sidechannel
+would allow for quicker, more efficient communication between processes access the same shard.
+
+An alternative to this architecture, that would obsolete the need for a sidechannel entirely, would be to consolidate
+interactions with Persist purely into `environmentd`, and allow it to instruct `clusterd` what work to perform explicitly.
+Some ideas have been floating that [move in this direction](https://materializeinc.slack.com/archives/C04LEA9D8BU/p16813495590089590).
+
+After much thought on this one, our determination is that, regardless of whether we make broader architectural changes,
+a Persist sidechannel is worth pursuing now, even if it is obsoleted and ripped out later. It would meaningfully benefit
+the product and our infra costs sooner rather than later, its scope is well contained, and it would not introduce any new
+complexity that would impede any larger reworkings.
+
+## Sidechannel Implementations
+
+While we have proposed an RPC-based implementation that flows through `environmentd`, these are the alternatives we've
+considered:
+
+### Controller RPCs
+
+We could push the RPCs into the Controller, and use the existing Adapter<->Compute|Storage interfaces to communicate diffs.
+This would avoid any complexity involved in setting up new connections and request handling, and allow the Controller to
+precisely specify which shards are needed by which processes, which is information that it already knows.
+
+To this point, we've rejected this option for several reasons: Persist sidechannel traffic could become a noisy neighbor
+(or by affected by noisy neighbors) in the existing request handling; pushing diff updates is purely an optimization while
+other messages in the Controller RPC stream are critical for correctness and overall operations; Persist internal details
+leak into the Controller in a way that weakens its abstraction.
+
+### Point-to-Point RPCs / Mesh
+
+We could alternatively set up the proposed gRPC service to connect `environmentd` and each `clusterd` to each other
+directly, creating a mesh. In this model, `environmentd` would no longer need to broadcast out changes -- each process
+would send diffs directly to the processes that need them.
+
+This approach is additionally interesting in that it would set Persist up to communicate other types of data as well,
+not just pub-sub of diffs. One hypothetical -- shard writers could cache recently-written blobs to S3 in memory or on
+disk, and other processes could request those blobs from the writer directly, rather than paying the costs of going to
+S3, which is likely to have much longer tail latencies.
+
+The downsides of this approach include the complexity overhead, as it introduces the novel pattern of intra-`clusterd`
+communication in a way that does not exist today, and the scaling limitations imposed by all-to-all communication.
```

Comment:
Besides an increased number of connections and opened sockets, what are the scaling limitations?